packages/flutter/lib/src/widgets/page_view.dart

```diff
@@ -82,21 +152,92 @@ class _PagePosition extends ScrollPosition {
 
   final int initialPage;
 
-  double get page => pixels / viewportDimension;
+  double get peek => _peek;
+  double _peek;
+  set peek(double newValue) {
+    if (_peek == newValue)
+      return;
+    final double oldPage = page;
+    _peek = newValue;
+    if (oldPage != null)
+      correctPixels(getPixelsFromPage(oldPage));
+  }
+
+  double getPageFromPixels(double pixels, double viewportDimension) {
+    return math.max(0.0, pixels) / math.max(1.0, viewportDimension - 2.0 * peek);
+  }
+
+  double getPixelsFromPage(double page) {
+    return page * math.max(0.0, viewportDimension - 2.0 * peek);
+  }
+
+  double get page => pixels == null ? null : getPageFromPixels(pixels, viewportDimension);
 
   @override
   bool applyViewportDimension(double viewportDimension) {
     final double oldViewportDimensions = this.viewportDimension;
     final bool result = super.applyViewportDimension(viewportDimension);
     final double oldPixels = pixels;
-    final double page = (oldPixels == null || oldViewportDimensions == 0.0) ? initialPage.toDouble() : oldPixels / oldViewportDimensions;
-    final double newPixels = page * viewportDimension;
+    final double page = (oldPixels == null || oldViewportDimensions == 0.0) ? initialPage.toDouble() : getPageFromPixels(oldPixels, oldViewportDimensions);
+    final double newPixels = getPixelsFromPage(page);
     if (newPixels != oldPixels) {
       correctPixels(newPixels);
       return false;
     }
     return result;
   }
+
+  @override
+  PageMetrics getMetrics() {
+    return new PageMetrics(
+      parent: super.getMetrics(),
+      page: page,
+    );
+  }
+}
+
+/// Scroll physics used by a [PageView].
+///
+/// These physics cause the page view to snap to page boundaries.
+class PageScrollPhysics extends ScrollPhysics {
+  /// Creates physics for a [PageView].
+  const PageScrollPhysics({ ScrollPhysics parent }) : super(parent);
+
+  @override
+  PageScrollPhysics applyTo(ScrollPhysics parent) => new PageScrollPhysics(parent: parent);
+
+  double _getPage(ScrollPosition position) {
```

Comment:
Does this work if peek > viewportDimension / 4, i.e. if 3 or more pages are visible after the pageview has settled?